asm/interrupt.s

```diff
@@ -0,0 +1,53 @@
+#include "ultra64/asm.h"
+#include "ultra64/r4300.h"
+#include "ultra64/thread.h"
+
+# assembler directives
+.set noat      # allow manual use of $at
+.set noreorder # don't insert nops after branches
+.set gp=64     # allow use of 64-bit general purpose registers
+
+.section .text
+
+.balign 16
+
+LEAF(__osDisableInt)
+    lui     $t2, %hi(__OSGlobalIntMask)
+    addiu   $t2, $t2, %lo(__OSGlobalIntMask)
+    lw      $t3, ($t2)
+    andi    $t3, $t3, SR_IMASK
+    mfc0    $t0, C0_SR
+    li      $at, ~SR_IE
+    and     $t1, $t0, $at
+    mtc0    $t1, C0_SR
+    andi    $v0, $t0, SR_IE
+    lw      $t0, ($t2)
+    andi    $t0, $t0, SR_IMASK
+    beq     $t0, $t3, No_Change_Global_Int
+     lui    $t2, %hi(__osRunningThread)
+    addiu   $t2, $t2, %lo(__osRunningThread)
+    lw      $t1, THREAD_SR($t2)
```

Comment:
:thinking: here,
`t2 = &__osRunningThread` is a `OSThread**`
so `lw $t1, THREAD_SR($t2)` does
`(OSThread*&__osRunningThread)->context.sr`
which doesn't make much sense to me
I would expect `__osRunningThread->context.sr`

`(OSThread*&__osRunningThread)->context.sr` is inside `osViModeMpalLan1`:
```
$ ./sym_info.py __osRunningThread
Symbol __osRunningThread (RAM: 0x8000AD50, ROM: 0x00B950, build/src/libultra/os/createthread.o)
```
`0x8000AD50 + THREAD_SR = 0x8000AE68`
```
$ ./sym_info.py 0x8000AE68
0x8000AE68 is at 0x18 bytes inside osViModeMpalLan1 (RAM 0x8000AE50, ROM 0xBA50, build/src/libultra/io/vimodempallan1.o)
```

is this stuff totally broken and if so how does it even work or what am I missing :cold_sweat:

Reply:
Yes this is another bug, the `addiu` should definitely be `lw` instead.
That said, I'm reasonably sure the branch is always taken, so the broken code will never run. The branch is essentially `(__OSGlobalIntMask & SR_IMASK) == (__OSGlobalIntMask & SR_IMASK)`
The only way this could ever fail is if some interrupt occurred and changed `__OSGlobalIntMask`, however interrupts are disabled pretty quickly so I don't think such great timing would occur usually. I'm not really sure what their goal was with this code, it's actually missing in earlier libultra versions.